src/main/core/work/task.rs

```diff
@@ -5,57 +5,35 @@ use atomic_refcell::AtomicRefCell;
 use crate::{
     core::worker::Worker,
     host::host::Host,
-    utility::{IsSend, IsSync},
+    utility::{IsSend, IsSync, Magic},
 };
 
 /// Mostly for interoperability with C APIs.
 /// In Rust code that doesn't need to interact with C, it may make more sense
 /// to directly use a `FnMut(&mut Host)` trait object.
 pub struct TaskRef {
     inner: Arc<AtomicRefCell<dyn FnMut(&mut Host) + Send + Sync>>,
-    // Runtime memory error checking to help catch errors that C code is prone
-    // to.  Can probably drop once C interop is removed, if we don't drop this
-    // type altogether.
-    #[cfg(debug_assertions)]
-    magic: u32,
+    magic: Magic<0xe0408897>,
 }
 
 impl TaskRef {
-    #[cfg(debug_assertions)]
-    const MAGIC: u32 = 0xe0408897;
-
     pub fn new<T: 'static + FnMut(&mut Host) + Send + Sync>(f: T) -> Self {
         Worker::increment_object_alloc_counter("TaskRef");
         Self {
             inner: Arc::new(AtomicRefCell::new(f)),
-            #[cfg(debug_assertions)]
-            magic: Self::MAGIC,
+            magic: Magic::new(),
         }
     }
 
     pub fn execute(&mut self, host: &mut Host) {
-        self.check_magic();
+        self.magic.debug_check();
         let mut inner = self.inner.borrow_mut();
         inner(host)
     }
-
-    fn clear_magic(&mut self) {
-        #[cfg(debug_assertions)]
-        unsafe {
-            std::ptr::write_volatile(&mut self.magic, 0)
-        };
-    }
-
-    fn check_magic(&self) {
-        #[cfg(debug_assertions)]
-        debug_assert!(self.magic == Self::MAGIC);
-    }
 }
 
 impl Drop for TaskRef {
     fn drop(&mut self) {
-        self.check_magic();
-        self.clear_magic();
         Worker::increment_object_dealloc_counter("TaskRef");
     }
 }
```

Comment:
I don't follow why this would be too late. Do you have an example where something would go wrong?

The `Magic` in this case protects the `TaskRef`, not the "Task". Nothing else should be able to access the `TaskRef` while `drop` is running, since it has a mutable reference to `self`.

More generally I'd like to avoid forcing an explicit `Drop` implementation for any struct that uses `Magic`. I started to write a comment that a `Magic` member should always be first so that it's de-init'd first, but it's not clear to me that it's needed (though it shouldn't hurt, either)

Reply:
If the C code called `taskref_drop()` on an already-dropped `TaskRef` (the equivalent of a double-free), a `Box` will be created pointing to an invalid `TaskRef`. The `Box` will then be dropped, which drops an invalid `TaskRef`. The `TaskRef` drop implementation will first try to drop an invalid `Arc` which would be bad. I think we want to check the magic value first before the drop implementation attempts to modify other invalid struct fields.

Reply:
Oh! I was thinking you were concerned about the placement of `clear_magic`, but you clearly said `check_magic`, which absolutely makes sense.

I am going to do it via field order and add a comment to `Magic` though. I'd like this and other such C objects not to have to all provide their own Drop impls. A follow-up PR adds an analagous helper for doing the counter tracking, which lets us get rid of this one.